Fix various memory leaks and mismatched delete

kadler opened the pull request:
The biggest leak is from ODBCConnection::CallProcedure, which calls
deleteColumns twice after calling SQLProcedures and SQLProcedureColumns.
This function does not delete any of the boundRow array members, only
the array itself. Likewise, it clears the storedRows vector, but does
not delete any of the pointers contained within. The code to properly do
this is in the clear method, but it also clears out the bound parameters
which we need to save once we call the actual procedure. This should
probably be changed so that is not needed, but for now we can
consolidate the necessary cleanup code form clear in to deleteColumns
and the inline the rest into the destructor (since clear is only called
there) along with a call to deleteColumns.

Additionally, make sure to delete other pointers passed to the various
async worker classes and fix "Mismatched free() / delete / delete []"
issues found by Valgrind in the process of investigating the leak.

Fixes #194

markdirish commented:
The bad news: Now that the 2.4.0 branch has been merged into `master`, there are conflicts that must be resolved.
The good news: That now means that these changes will be included in 2.4.0

kadler commented:
I've rebased the changes and due to the numerous changes from the 2.4.0 branch, I basically re-ported the changes. There were some redundant if checks I removed, so the indentation has changed on some of the for loops, so viewing the diff without whitespace changes makes things clearer.

markdirish reviewed (approved):
LGTM
